polylaue/model/editable.py

```diff
@@ -0,0 +1,184 @@
+# Copyright © 2024, UChicago Argonne, LLC. See "LICENSE" for full details.
+
+from typing import (
+    Any,
+    Callable,
+    Literal,
+    NotRequired,
+    Sequence,
+    TypedDict,
+    Union,
+)
+
+from pathlib import Path
+
+from polylaue.model.serializable import Serializable, ValidationError
+
+Number = Union[int, float]
+ParameterType = Literal[
+    "string", "integer", "float", "boolean", "file", "folder", "tuple"
+]
+ParameterValue = Union[str, Number]
+ParameterOptions = Sequence[ParameterValue]
+
+
+# NotRequired requires python >=3.11
+class ParameterDescription(TypedDict):
+    type: ParameterType
+    label: str
+    required: NotRequired[bool]
+    subtype: NotRequired[ParameterType]
+    length: NotRequired[int]
+    default: NotRequired[ParameterValue]
+    options: NotRequired[ParameterOptions]
+    min: NotRequired[Number]
+    max: NotRequired[Number]
+
+
+ParameterValidator = Callable[
+    [Any, ParameterDescription, dict[str, ParameterValue], "Editable"], None
+]
+
+
+def default_string_validator(value, description: ParameterDescription, *args):
```

Comment:
What are the `*args` there for?

Reply:
While the feature is not used in this PR, I laid out the ground work to have dynamic custom validators. These validators take not only the value of one parameter, but they also are provided the value of all other parameters as well as a reference to the `Editable` that they are being applied to. This is the full signature:

```python
def validator(value: Any, description: ParameterDescription, params: dict[str, ParameterValue], editable: "Editable") -> None:
    pass
```

And they are invoked in the default implementation of `Editable.validate_parameters()`:
```python
class Editable(Serializable):
    ...
    def validate_parameters(self, params):
        ...
            validator = self.custom_validators.get(k)

            if validator is None:
                validator = DEFAULT_VALIDATORS.get(
                    description['type'], noop_validator
                )

            validator(value, description, params, self)
```

So in the case of the default validators, we only use the value and the description, but we capture the other parameters passed to them with `*args`.

Reply:
We might want to make optional arguments into keyword arguments and use `**kwargs` instead of `*args`.